scripts/azureml-assets/azureml/assets/release/group_test.py

```diff
@@ -40,42 +80,91 @@
-            print(f'Loading test job {job}')
-            test_job = azure.ai.ml.load_job(tests_dir / job_data['job'])
-            print(test_job)
-            print(f'Running test job {job}')
-            test_job = ml_client.jobs.create_or_update(test_job)
-            print(f'Submitted test job {job}')
-            print(f"Job id: {test_job.id}")
-            submitted_job_list.append(test_job)
+            if 'pytest_job' in job_data:
+                pytest_jobs[tests_dir / job_data['pytest_job']] = job_data["assets"]
+            else:
+                if 'pre' in job_data:
+                    logger.print(f"Running pre script for {job}")
+                    proc = check_call(f"python3 {tests_dir / job_data['pre']}", env=my_env, shell=True)
+                print(f'Loading test job {job}')
+                try:
+                    test_job = azure.ai.ml.load_job(tests_dir / job_data['job'])
+                    logger.print(f'Running test job {job}')
+                    test_job = ml_client.jobs.create_or_update(test_job)
+                except Exception as ex:
+                    logger.log_warning(
+                        f"catch error submitting {job_data['job']} with exception {ex}")
+                    failed_jobs.append(job)
+                    continue
+                test_coverage[test_job] = job_data.get("assets", [])
+                logger.print(f'Submitted test job {job}')
+                logger.print(f"Job id: {test_job.id}")
+                submitted_job_list.append(test_job)
+
+    # Run pytest jobs
+    submitted_pytest_jobs = run_pytest_jobs(pytest_jobs, my_env)
 
+    # Process pytest results
+    # while submitted_pytest_jobs:
+    #    for job in submitted_pytest_jobs.copy().keys():
+    #        if job.done():
+    #            if job.result() == 0:
+    #                succeeded_jobs.append(job)
+    #                covered_assets.extend(submitted_pytest_jobs[job])
+    #            else:
+    #                failed_jobs.append(job)
+    #            del submitted_pytest_jobs[job]
+
+    for job in submitted_pytest_jobs.keys():
+        if job.result() == 0:
+            succeeded_jobs.append(job)
+            covered_assets.extend(submitted_pytest_jobs[job])
+        else:
+            failed_jobs.append(job)
+            
     # TO-DO: job post and group post scripts will be run after all jobs are completed
 
+    # process pipeline jobs results
     while submitted_job_list:
         for job in submitted_job_list:
             returned_job = ml_client.jobs.get(job.name)
             if returned_job.status == "Completed":
                 succeeded_jobs.append(returned_job.display_name)
                 submitted_job_list.remove(job)
+                covered_assets.extend(test_coverage.get(job, []))
             elif returned_job.status == "Failed" or returned_job.status == "Cancelled":
                 failed_jobs.append(returned_job.display_name)
                 submitted_job_list.remove(job)
```

Comment:
This is simpler. Also, are there other job statuses than the three handled below? There's no final `else` in the `if`/`else` below. Or can you handle any job whose status is not "Completed" as failed?
```suggestion
    for job in submitted_job_list:
        returned_job = ml_client.jobs.get(job.name)
        if returned_job.status == "Completed":
            succeeded_jobs.append(returned_job.display_name)
            covered_assets.extend(test_coverage.get(job, []))
        elif returned_job.status in ["Failed", "Cancelled"]:
            failed_jobs.append(returned_job.display_name)
```

Reply:
There are only these 3 statuses that are considered the final statuses of a job.